File: mahjong/analyzer.py
from __future__ import annotations

from dataclasses import dataclass

from mahjong.shanten import normal_shanten
from mahjong.tiles import tile_name, tile_index, validate_counts


GOOD_WAIT_THRESHOLD = 5
SUPER_GOOD_WAIT_THRESHOLD = 9


@dataclass(frozen=True)
class TenpaiDetail:
    draw: str
    draw_tiles: int
    winning_types: int
    winning_tiles: int
    winning: tuple[str, ...]
    tenpai_discards: tuple[str, ...]

    @property
    def is_good_shape(self) -> bool:
        return self.winning_tiles >= GOOD_WAIT_THRESHOLD

    @property
    def is_super_good_shape(self) -> bool:
        return self.winning_tiles >= SUPER_GOOD_WAIT_THRESHOLD


@dataclass(frozen=True)
class DiscardAnalysis:
    discard: str
    after_discard_shanten: int
    is_iishanten: bool
    ukeire_types: int
    ukeire_tiles: int
    ukeire: tuple[str, ...]
    good_shape_types: int
    good_shape_tiles: int
    good_shape_rate: float
    super_good_shape_types: int
    super_good_shape_tiles: int
    super_good_shape_rate: float
    tenpai_details: tuple[TenpaiDetail, ...]

    @property
    def sort_key(self) -> tuple[int, int, int]:
        return (-self.after_discard_shanten, self.ukeire_types, self.ukeire_tiles)

    @property
    def progress_shanten(self) -> int:
        return self.after_discard_shanten - 1


def analyze_discards(counts: tuple[int, ...] | list[int]) -> list[DiscardAnalysis]:
    """Compare all unique discard candidates for a 14-tile hand."""
    validate_counts(counts, expected_total=14)
    counts_list = list(counts)
    results: list[DiscardAnalysis] = []

    for discard_index, count in enumerate(counts_list):
        if count == 0:
            continue

        after_discard = counts_list.copy()
        after_discard[discard_index] -= 1
        shanten = normal_shanten(tuple(after_discard))
        ukeire = _calculate_ukeire(after_discard, shanten)
        tenpai_details = _calculate_tenpai_details(after_discard, ukeire) if shanten == 1 else ()
        good_shape_tiles = sum(detail.draw_tiles for detail in tenpai_details if detail.is_good_shape)
        super_good_shape_tiles = sum(detail.draw_tiles for detail in tenpai_details if detail.is_super_good_shape)
        ukeire_tiles = sum(4 - after_discard[tile_index(tile)] for tile in ukeire)

        results.append(
            DiscardAnalysis(
                discard=tile_name(discard_index),
                after_discard_shanten=shanten,
                is_iishanten=shanten == 1,
                ukeire_types=len(ukeire),
                ukeire_tiles=ukeire_tiles,
                ukeire=ukeire,
                good_shape_types=sum(1 for detail in tenpai_details if detail.is_good_shape),
                good_shape_tiles=good_shape_tiles,
                good_shape_rate=_safe_rate(good_shape_tiles, ukeire_tiles),
                super_good_shape_types=sum(1 for detail in tenpai_details if detail.is_super_good_shape),
                super_good_shape_tiles=super_good_shape_tiles,
                super_good_shape_rate=_safe_rate(super_good_shape_tiles, ukeire_tiles),
                tenpai_details=tenpai_details,
            )
        )

    return sorted(results, key=lambda item: item.sort_key, reverse=True)


def best_discards(results: list[DiscardAnalysis]) -> list[DiscardAnalysis]:
    eligible = _best_shanten_results(results)
    if not eligible:
        return []
    best_key = max((result.ukeire_types, result.ukeire_tiles) for result in eligible)
    return [result for result in eligible if (result.ukeire_types, result.ukeire_tiles) == best_key]


def best_discards_for_review(results: list[DiscardAnalysis]) -> list[DiscardAnalysis]:
    """Return correct answers for review mode.

    Review mode prioritizes physical ukeire tile count. If the ukeire tile
    count ties, the higher good-shape tile count wins. Equal good-shape count
    remains a multi-answer correct result.
    """
    eligible = _best_shanten_results(results)
    if not eligible:
        return []
    best_key = max((result.ukeire_tiles, result.good_shape_tiles) for result in eligible)
    return [result for result in eligible if (result.ukeire_tiles, result.good_shape_tiles) == best_key]


def _best_shanten_results(results: list[DiscardAnalysis]) -> list[DiscardAnalysis]:
    if not results:
        return []
    best_shanten = min(result.after_discard_shanten for result in results)
    return [result for result in results if result.after_discard_shanten == best_shanten]


def _calculate_ukeire(after_discard: list[int], shanten: int) -> tuple[str, ...]:
    target_shanten = shanten - 1
    ukeire: list[str] = []
    for draw_index in range(34):
        if after_discard[draw_index] >= 4:
            continue
        drawn = after_discard.copy()
        drawn[draw_index] += 1
        if normal_shanten(tuple(drawn)) <= target_shanten:
            ukeire.append(tile_name(draw_index))
    return tuple(ukeire)


def _calculate_tenpai_details(after_discard: list[int], ukeire: tuple[str, ...]) -> tuple[TenpaiDetail, ...]:
    details: list[TenpaiDetail] = []
    for draw in ukeire:
        draw_index = tile_index(draw)
        drawn = after_discard.copy()
        drawn[draw_index] += 1

        best_winning: tuple[str, ...] = ()
        best_discards: list[str] = []
        best_winning_tiles = -1

        for discard_index, count in enumerate(drawn):
            if count == 0:
                continue
            tenpai = drawn.copy()
            tenpai[discard_index] -= 1
            if normal_shanten(tuple(tenpai)) != 0:
                continue

            winning = _calculate_winning_tiles(tenpai)
            winning_tiles = sum(4 - tenpai[tile_index(tile)] for tile in winning)
            if winning_tiles > best_winning_tiles:
                best_winning = winning
                best_discards = [tile_name(discard_index)]
                best_winning_tiles = winning_tiles
            elif winning_tiles == best_winning_tiles:
                best_discards.append(tile_name(discard_index))

        details.append(
            TenpaiDetail(
                draw=draw,
                draw_tiles=4 - after_discard[draw_index],
                winning_types=len(best_winning),
                winning_tiles=max(best_winning_tiles, 0),
                winning=best_winning,
                tenpai_discards=tuple(best_discards),
            )
        )

    return tuple(details)


def _calculate_winning_tiles(tenpai: list[int]) -> tuple[str, ...]:
    winning: list[str] = []
    for draw_index in range(34):
        if tenpai[draw_index] >= 4:
            continue
        won = tenpai.copy()
        won[draw_index] += 1
        if normal_shanten(tuple(won)) == -1:
            winning.append(tile_name(draw_index))
    return tuple(winning)


def _safe_rate(numerator: int, denominator: int) -> float:
    if denominator == 0:
        return 0.0
    return numerator / denominator

File: packages/mahjong-core/src/analyzer.ts
import { normalShanten } from "./shanten";
import { type Counts34, type Tile, tileIndex, tileName, validateCounts } from "./tiles";

export const GOOD_WAIT_THRESHOLD = 5;
export const SUPER_GOOD_WAIT_THRESHOLD = 9;

export interface TenpaiDetail {
  draw: Tile;
  drawTiles: number;
  winningTypes: number;
  winningTiles: number;
  winning: Tile[];
  tenpaiDiscards: Tile[];
}

export interface DiscardAnalysis {
  discard: Tile;
  afterDiscardShanten: number;
  isIishanten: boolean;
  ukeireTypes: number;
  ukeireTiles: number;
  ukeire: Tile[];
  goodShapeTypes: number;
  goodShapeTiles: number;
  goodShapeRate: number;
  superGoodShapeTypes: number;
  superGoodShapeTiles: number;
  superGoodShapeRate: number;
  tenpaiDetails: TenpaiDetail[];
}

export function analyzeDiscards(counts: Counts34): DiscardAnalysis[] {
  validateCounts(counts, 14);
  const results: DiscardAnalysis[] = [];

  counts.forEach((count, discardIndex) => {
    if (count === 0) return;

    const afterDiscard = counts.slice();
    afterDiscard[discardIndex] -= 1;
    const shanten = normalShanten(afterDiscard);
    const ukeire = calculateUkeire(afterDiscard, shanten);
    const tenpaiDetails = shanten === 1 ? calculateTenpaiDetails(afterDiscard, ukeire) : [];
    const goodShapeTiles = tenpaiDetails.filter((detail) => detail.winningTiles >= GOOD_WAIT_THRESHOLD).reduce((sum, detail) => sum + detail.drawTiles, 0);
    const superGoodShapeTiles = tenpaiDetails.filter((detail) => detail.winningTiles >= SUPER_GOOD_WAIT_THRESHOLD).reduce((sum, detail) => sum + detail.drawTiles, 0);
    const ukeireTiles = ukeire.reduce((sum, tile) => sum + 4 - afterDiscard[tileIndex(tile)]!, 0);

    results.push({
      discard: tileName(discardIndex),
      afterDiscardShanten: shanten,
      isIishanten: shanten === 1,
      ukeireTypes: ukeire.length,
      ukeireTiles,
      ukeire,
      goodShapeTypes: tenpaiDetails.filter((detail) => detail.winningTiles >= GOOD_WAIT_THRESHOLD).length,
      goodShapeTiles,
      goodShapeRate: safeRate(goodShapeTiles, ukeireTiles),
      superGoodShapeTypes: tenpaiDetails.filter((detail) => detail.winningTiles >= SUPER_GOOD_WAIT_THRESHOLD).length,
      superGoodShapeTiles,
      superGoodShapeRate: safeRate(superGoodShapeTiles, ukeireTiles),
      tenpaiDetails,
    });
  });

  return results.sort((a, b) => sortValue(b) - sortValue(a));
}

export function bestDiscardsForReview(results: DiscardAnalysis[]): DiscardAnalysis[] {
  if (results.length === 0) return [];
  const bestShanten = Math.min(...results.map((result) => result.afterDiscardShanten));
  const eligible = results.filter((result) => result.afterDiscardShanten === bestShanten);
  const bestTiles = Math.max(...eligible.map((result) => result.ukeireTiles));
  const bestGoodShape = Math.max(...eligible.filter((result) => result.ukeireTiles === bestTiles).map((result) => result.goodShapeTiles));
  return eligible.filter((result) => result.ukeireTiles === bestTiles && result.goodShapeTiles === bestGoodShape);
}

function calculateUkeire(afterDiscard: Counts34, shanten: number): Tile[] {
  const targetShanten = shanten - 1;
  const ukeire: Tile[] = [];
  for (let drawIndex = 0; drawIndex < 34; drawIndex += 1) {
    if (afterDiscard[drawIndex]! >= 4) continue;
    const drawn = afterDiscard.slice();
    drawn[drawIndex] += 1;
    if (normalShanten(drawn) <= targetShanten) {
      ukeire.push(tileName(drawIndex));
    }
  }
  return ukeire;
}

function calculateTenpaiDetails(afterDiscard: Counts34, ukeire: Tile[]): TenpaiDetail[] {
  return ukeire.map((draw) => {
    const drawIndex = tileIndex(draw);
    const drawn = afterDiscard.slice();
    drawn[drawIndex] += 1;

    let bestWinning: Tile[] = [];
    let bestDiscards: Tile[] = [];
    let bestWinningTiles = -1;

    drawn.forEach((count, discardIndex) => {
      if (count === 0) return;
      const tenpai = drawn.slice();
      tenpai[discardIndex] -= 1;
      if (normalShanten(tenpai) !== 0) return;

      const winning = calculateWinningTiles(tenpai);
      const winningTiles = winning.reduce((sum, tile) => sum + 4 - tenpai[tileIndex(tile)]!, 0);
      if (winningTiles > bestWinningTiles) {
        bestWinning = winning;
        bestDiscards = [tileName(discardIndex)];
        bestWinningTiles = winningTiles;
      } else if (winningTiles === bestWinningTiles) {
        bestDiscards.push(tileName(discardIndex));
      }
    });

    return {
      draw,
      drawTiles: 4 - afterDiscard[drawIndex]!,
      winningTypes: bestWinning.length,
      winningTiles: Math.max(bestWinningTiles, 0),
      winning: bestWinning,
      tenpaiDiscards: bestDiscards,
    };
  });
}

function calculateWinningTiles(tenpai: Counts34): Tile[] {
  const winning: Tile[] = [];
  for (let drawIndex = 0; drawIndex < 34; drawIndex += 1) {
    if (tenpai[drawIndex]! >= 4) continue;
    const won = tenpai.slice();
    won[drawIndex] += 1;
    if (normalShanten(won) === -1) {
      winning.push(tileName(drawIndex));
    }
  }
  return winning;
}

function sortValue(result: DiscardAnalysis): number {
  return -result.afterDiscardShanten * 100000 + result.ukeireTypes * 1000 + result.ukeireTiles;
}

function safeRate(numerator: number, denominator: number): number {
  return denominator === 0 ? 0 : numerator / denominator;
}
